### DISCUSSION TITLE
Compile/Build from strings instead of files

### DISCUSSION THREAD

**Discussion by dashtinejad** [11ty/eleventy#2871] [Enhancement Queue]
Hi,

I've been looking to compile a string via eleventy programmatic API, but I couldn't find anything. I guess it's not implemented in this way.

So the idea here is to be able to compile a template, directly from string, instead of reading it from a file.

And the motivation behind it, is that the files are stored in a database, instead of files. So I would like to get them on the fly and be able to compile them via eleventy.

I have something like this in mind:

```js
let template = "Hello {{ name }}"; // is fetched from database
let model = { name: 'Eleventy' }; // is fetched from database

let elev = new Eleventy();
let html = await elev.compile('nunjucks', template, model);

console.log(html) // Hello Eleventy
```

Is it possible to implement? Or is it against the philosophy behind eleventy? I believe eleventy has a great engine for compiling template languages into HTML, but it's coupled to files at the moment.

Glad to hear your opinions on this. Thanks!

**Comment by dashtinejad** [comment]
Any thoughts?

**Reply by Ryuno-Ki** [reply]
Sounds like an use case for [virtual templates](https://www.11ty.dev/docs/virtual-templates/).
